### ISSUE TITLE
Disable refunds and burn all gas

### ISSUE THREAD

**Issue by nearmax** [nearprotocol/nearcore#3161]
After a brief discussion with @SkidanovAlex I think this issue is severe enough to be promoted to Phase 1. Also, I think https://github.com/nearprotocol/NEPs/issues/104 is not framing the right problem: the problem is not that it is not expensive enough to perform shard congestion, the problem that it is possible in the first place. It should not be possible to disable transaction processing of a shard for 6mins without losing a large stake in the system, it shouldn't be merely expensive.

# Problem description

Suppose, there is a contract that burns 300Tgas during its execution. Suppose I create 200 transactions that call this contract and submit it asynchronously from multiple accounts so that they end up in the same block. All 200 transactions are going to be admitted into a single block, because converting a single function call transaction to a receipt only costs ~2.5Tgas. Unfortunately only 2 such function calls can be processed per block, which means for the next 100 blocks the shard will be doing nothing but processing transactions. Resulting in almost 2min downtime for the clients that are using our blockchain.

The cost of a single such attack is 60NEAR, but the attacker then can repeat this attack after delayed receipts are processed. 

# Broken invariant

The root of the problem is that we are breaking the following invariant:
> In a given time interval T it is possible to submit transactions with attached gas that can potentially result in more CPU computation that can be processed in a time interval of the same length T.

It is a flow problem -- if source produces more flow than sink can accept then it will accumulate somewhere.

# Heuristics

As long, as the variant is broken, no amount of heuristics can fix it. Examples of broken heuristics that will not work:
* Increasing gas price based on how long the shard has been congested;
* Increasing gas price based on how much the shard has been congested, e.g. the size of the delayed receipts queue;
Contr-argument to the above heuristics:
* Since attacker can perform attack within one block they are not affected by the price change, so the attack is already viable. Attacker can perform this attack at the random times of the day when the gas price is low. For instance, they can cause 1h total downtime per day, by paying 3600NEAR, which is negligible compared to the damadge it can cause. And this attack is completely agnostic to how much the gas price hikes after the block that admits 200 transactions.

* **Receipt priority queue based on gas price.** Every function call has gas price attached to it. Receipts in the delayed queue with higher gas price are processed first.
Contr-argument:
* It does not solve the fundamental issue -- the congestion is still possible. The 2 min congestion will have exactly the same price, and the 
* The second-order effects are not studied. It might open a full surface of attack vectors and manipulations that we need to argue about. It is not the first time when we introduce a seemingly simple change with complex second order effects that later create significant difficulties. For example, state staking seemed to be simple at first, until we figured out that now some contracts need attached tokens, and can be locked by state exhaustion;
* Our DevX and UX will need to be significantly more complex. Now developers and partners need to think how to implement the mechanics that would boost their receipts that get stuck. Some of them might need to include additional UI elements and educate their users on "stuck" receipts the way metamask has special functionality to boost transactions. This significantly degrades our UX;
* Our Applayer would need to be reworked. Components like bridge would need to have special complex logic to track all receipts produced by the given transaction and unstuck all of them. Wallet would need to have special UI elements and mechanics.

# Solution
It is clear we need to unbreak the invariant. The only way to unbreak the variant is to make sure that each block only contains transactions and receipts with prepaid gas that is cumulatively less than the block capacity.

Unfortunately, there is no incentive for users not to set too high prepaid gas, since we reimburse all unused gas. Which means people can fill in blocks with transactions that have 300Tgas attached but burn only 5Tgas, preventing everyone else from using the blockchain at a very low cost (0.015NEAR per second or 54NEAR per hour). To make sure users do not overestimate the prepaid gas, we need to burn it.
Advantages:
* Congestion problem is solved in its entirety;
* We don't need different gas price between the shards;
* Without it the prepaid gas is some magic number that users don't care about and set to max, which is a weird concept on its own;
* There are no refunds, which means there are 30% fewer receipts for function call transactions:
  * As the result contract call TPS is higher;
  * Contract call finality is 50% faster.
* Receipts cannot be delayed for infinite amount of time. 

Disadvantages:
* The DevX becomes a pain. You need to precisely estimate gas for cross-contract calls to avoid overcharging the user, but also if you underestimate, the chain of calls might fail at unexpected state.
   * Contr-argument: you don't need to precisely estimate the gas, because our fees are cheap. Besides, Ethereum users have to ballpark gasLimit anyway, because miners do not like large gasLimit transactions (see https://ethereum.stackexchange.com/questions/70723/is-there-a-drawback-of-setting-a-too-high-gas-limit-for-a-transaction). So this is not a new DevX and costs are low;
* It doesn't fully solve the shard congesting, because multiple shards might route receipts to one shard and still create a delayed queue. But this delayed queue can only be blocked by N blocks (where N is the number of shards) per block.
  * Contr-argument: It fully solves shard congestion, because the total source capacity of the blockchain is equal to its total sink. All shard congestions are temporary and resolvable through resharding.

----

I suggest we go with the full-burn solution. We know it is bullet-proof, if later we come up with a scheme that allows refunds we implement it through upgrades. Doing it the other way around -- turning off the upgrades post Phase 1 launch is going to be significantly more painful to our users.

**Comment by bowenwang1996** [comment]
> Unfortunately, there is no incentive for users not to set too high prepaid gas, since we reimburse all unused gas. Which means people can fill in blocks with transactions that have 300Tgas attached but burn only 5Tgas

I don't think that is the case. When we process receipts, we look at burnt gas to determine the limit. This means that if there are a lot of transactions that have 300Tgas attached but only uses 5Tgas, we will be able to process all receipts in one block. I agree that it is a problem if each of the transaction actually burns 300Tgas, but if that is the case, I don't think what proposed here helps either.

**Comment by nearmax** [comment]
> > Unfortunately, there is no incentive for users not to set too high prepaid gas, since we reimburse all unused gas. Which means people can fill in blocks with transactions that have 300Tgas attached but burn only 5Tgas
> 
> I don't think that is the case. When we process receipts, we look at burnt gas to determine the limit. This means that if there are a lot of transactions that have 300Tgas attached but only uses 5Tgas, we will be able to process all receipts in one block. I agree that it is a problem if each of the transaction actually burns 300Tgas, but if that is the case, I don't think what proposed here helps either.

The argument that you quoted explains why users don't have incentive to care about the prepaid gas. The contr-argument: "This means that if there are a lot of transactions that have 300Tgas attached but only uses 5Tgas, we will be able to process all receipts in one block." does not argue that users will care about the prepaid gas. Here is an example: Imagine we've been running network for 2 months, users do not care about prepaid gas, because it is refunded, as a bonus all transactions are processed without delay. Then malicious user comes in and exploits the fact that they can stall the system for 2 mins. What I am arguing is "we need to unbreak the variant" => "we need to fill block using prepaid gas" => "prepaid gas need to be close to used gas or else blocks will be underutilized" => "users need to care about prepaid gas" => "we need to burn all prepaid gas". What are trying to disprove "user don't care about the gas" => "the blocks are congested" which is not what I am saying.

**Comment by bowenwang1996** [comment]
I don't think I understand. How is burning all prepaid gas different from a transaction that uses all prepaid gas in the current setting?

**Comment by nearmax** [comment]
> I don't think I understand. How is burning all prepaid gas different from a transaction that uses all prepaid gas in the current setting?

I don't understand your question.

**Comment by bowenwang1996** [comment]
I think I understand your argument now. Let's say that we burn all prepaid gas. Now if an attacker wants to execute the same attack they have to saturate every block with transactions that attach a lot of gas. But the overall cost is the same as the attack in the current system because we charge receipts at the gas price of the block in which they are processed. Is your argument that it is much harder for an attacker to continuously saturate blocks with transactions?

**Comment by nearmax** [comment]
> I think I understand your argument now. Let's say that we burn all prepaid gas. Now if an attacker wants to execute the same attack they have to saturate every block with transactions that attach a lot of gas. But the overall cost is the same as the attack in the current system because we charge receipts at the gas price of the block in which they are processed. Is your argument that it is much harder for an attacker to continuously saturate blocks with transactions?

The argument is that there won't be a difference between the attacker and the user -- if user is willing to saturate the block and pay for it it does not matter whether they saturate it with useful computation, useless computation, or they underutilize the block but still pay for the unused CPU.

**Comment by bowenwang1996** [comment]
> they underutilize the block but still pay for the unused CPU.

It seems that you suggest that in the current system attacker can abuse without paying for unused CPU. I don't see why that is the case. As I mentioned before, the number of receipts we process depends on their burnt gas, not used gas. So I don't see how the attacker get away without paying for unused CPU.

**Comment by nearmax** [comment]
In the current system attacker can create a congestion for 2 mins.
In a system where all prepaid gas is burnt and blocks are filled in based on prepaid gas, and not just fees, it is not possible.

In the current system the attack described in the issue is possible, with the proposed modification it is not possible.

**Comment by bowenwang1996** [comment]
> In a system where blocks are filled in based on prepaid gas rather than fees the attacker can abuse without paying for CPU, therefore if we fill in based on prepaid gas we also need to burn all the prepaid gas.

But that is not the case. When we process receipts the block is filled based on burnt gas, not prepaid gas. So we can process a lot of receipts with 300T prepaid gas but 5T burnt gas.

**Comment by evgenykuzyakov** [comment]
Has to go to a NEP
